src/lib/consensus/proof_of_stake.ml

```diff
@@ -2899,25 +2972,38 @@ module Hooks = struct
         `Check_again epoch_end_time
 
   let frontier_root_transition (prev : Consensus_state.Value.t)
-      (next : Consensus_state.Value.t) ~local_state ~snarked_ledger =
+      (next : Consensus_state.Value.t) ~(local_state : Local_state.t)
+      ~snarked_ledger =
     if
       not
         (Epoch.equal
            (Consensus_state.curr_epoch prev)
            (Consensus_state.curr_epoch next))
     then (
-      let delegatee_table =
-        compute_delegatee_table
-          (Local_state.current_block_production_keys local_state)
-          ~iter_accounts:(fun f ->
-            Coda_base.Ledger.Any_ledger.M.iteri snarked_ledger ~f )
-      in
-      let ledger = Coda_base.Sparse_ledger.of_any_ledger snarked_ledger in
-      let epoch_snapshot = {Local_state.Snapshot.delegatee_table; ledger} in
       !local_state.last_epoch_delegatee_table
       <- Some !local_state.staking_epoch_snapshot.delegatee_table ;
-      !local_state.staking_epoch_snapshot <- !local_state.next_epoch_snapshot ;
-      !local_state.next_epoch_snapshot <- epoch_snapshot )
+      Coda_base.Ledger.Db.close !local_state.staking_epoch_snapshot.ledger ;
+      File_system.rmrf !local_state.staking_epoch_snapshot.location ;
+      !local_state.staking_epoch_snapshot
+      <- { ledger=
+             Coda_base.Ledger.Db.create_checkpoint
+               !local_state.next_epoch_snapshot.ledger
+               ~directory_name:!local_state.staking_epoch_snapshot.location ()
+         ; merkle_root= !local_state.next_epoch_snapshot.merkle_root
+         ; delegatee_table= !local_state.next_epoch_snapshot.delegatee_table
+         ; location= !local_state.staking_epoch_snapshot.location } ;
+      Coda_base.Ledger.Db.close !local_state.next_epoch_snapshot.ledger ;
```

Comment:
This method of migrating the epoch ledgers involves unnecessary IO. Specifically, it copies the next epoch snapshot to the staking epoch snapshot instead of just shifting pointers to where the staking epoch snapshot is stored. Instead, I would recommend doing something like this: keep 2 UUIDs for each of the snapshots. Store these 2 UUIDs both in memory and in a file so that you know which UUID corresponds to which snapshot when you load from disk. Now you can redescribe the epoch snapshot migration as follows:

1) delete the staking epoch snapshot's database
2) allocate a new UUID and new database
3) transfer the contents of the root snarked ledger to the new database
4) move the staking epoch snapshot to point to the UUID of the next epoch snapshot, set the next epoch snapshot to point to the new UUID, and update the pointers on the filesystem